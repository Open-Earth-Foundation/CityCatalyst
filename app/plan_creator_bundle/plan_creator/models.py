from pydantic import BaseModel, Field
from typing import Optional, List, Dict

# --- Request models ---


class CityContextData(BaseModel):
    locode: str = Field(..., min_length=1, description="UN/LOCODE identifier")
    populationSize: Optional[int] = Field(
        default=None, ge=0, description="Population size of the city"
    )


class CityEmissionsData(BaseModel):
    stationaryEnergyEmissions: Optional[float] = Field(
        default=None, ge=0, description="Stationary energy emissions"
    )
    transportationEmissions: Optional[float] = Field(
        default=None, ge=0, description="Transportation emissions"
    )
    wasteEmissions: Optional[float] = Field(
        default=None, ge=0, description="Waste emissions"
    )
    ippuEmissions: Optional[float] = Field(
        default=None, ge=0, description="Industrial processes and product use emissions"
    )
    afoluEmissions: Optional[float] = Field(
        default=None,
        ge=0,
        description="Agriculture, forestry, and other land use emissions",
    )


class CityData(BaseModel):
    cityContextData: CityContextData
    cityEmissionsData: CityEmissionsData


class PlanRequest(BaseModel):
    cityData: CityData
    actionId: str = Field(..., min_length=1, description="Action ID")
    language: str = Field(
        default="en", min_length=2, max_length=2, description="ISO Language code"
    )


# --- Response models ---


class StartPlanCreationResponse(BaseModel):
    taskId: str
    status: str


class CheckProgressResponse(BaseModel):
    status: str
    error: Optional[str] = None


### WIP for the plan creation response ###


class PlanMetadata(BaseModel):
    locode: str
    cityName: str
    actionId: str
    actionName: str


class PlanContent(BaseModel):
    title: str
    description: str
    subactions: List[dict] = Field(
        description="List of subactions with their descriptions"
    )
    municipalInstitutions: List[dict] = Field(
        description="List of institutions with their roles and sources"
    )
    milestones: List[dict] = Field(
        description="List of milestones with their descriptions"
    )
    merIndicators: List[dict] = Field(
        description="List of MER indicators with their descriptions"
    )
    climateRisks: List[dict] = Field(
        description="List of climate risks with their descriptions"
    )
    mitigationSectors: List[dict] = Field(
        description="List of mitigation sectors with their descriptions"
    )


class Plan(BaseModel):
    metadata: PlanMetadata
    content: Dict[str, PlanContent] = Field(
        description="Plan content in different languages, keyed by language code (e.g. 'en', 'es')"
    )
