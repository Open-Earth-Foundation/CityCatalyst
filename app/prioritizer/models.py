from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum

# --- Enums ---


class PrioritizationType(str, Enum):
    MITIGATION = "mitigation"  # Set to mitigation to only run prioritization for mitigation actions
    ADAPTATION = "adaptation"  # Set to adaptation to only run prioritization for adaptation actions
    BOTH = "both"  # Set to both to run prioritization for both mitigation and adaptation actions


# --- Request models ---


class CityContextData(BaseModel):
    locode: str = Field(..., min_length=1, description="UN/LOCODE identifier")
    populationSize: Optional[int] = Field(
        default=None, ge=0, description="Population size of the city"
    )


class CityEmissionsData(BaseModel):
    stationaryEnergyEmissions: Optional[float] = Field(
        default=None, ge=0, description="Stationary energy emissions"
    )
    transportationEmissions: Optional[float] = Field(
        default=None, ge=0, description="Transportation emissions"
    )
    wasteEmissions: Optional[float] = Field(
        default=None, ge=0, description="Waste emissions"
    )
    ippuEmissions: Optional[float] = Field(
        default=None, ge=0, description="Industrial processes and product use emissions"
    )
    afoluEmissions: Optional[float] = Field(
        default=None,
        ge=0,
        description="Agriculture, forestry, and other land use emissions",
    )


class CityData(BaseModel):
    cityContextData: CityContextData
    cityEmissionsData: CityEmissionsData


class PrioritizerRequest(BaseModel):
    cityData: CityData
    prioritizationType: PrioritizationType = Field(
        default=PrioritizationType.BOTH,
        description="Type of actions to prioritize: mitigation, adaptation, or both",
    )


class PrioritizerRequestBulk(BaseModel):
    cityDataList: List[CityData]
    prioritizationType: PrioritizationType = Field(
        default=PrioritizationType.BOTH,
        description="Type of actions to prioritize: mitigation, adaptation, or both",
    )


# --- Response models ---


class StartPrioritizationResponse(BaseModel):
    taskId: str
    status: str


class CheckProgressResponse(BaseModel):
    status: str
    error: Optional[str] = None


class MetaData(BaseModel):
    locode: str
    rankedDate: datetime


class Explanation(BaseModel):
    en: str
    es: str
    pt: str
    # Add all the languages required here and defined in api.py LANGUAGES constant


class RankedAction(BaseModel):
    actionId: str
    rank: int
    explanation: Optional[Explanation] = None


class PrioritizerResponse(BaseModel):
    metadata: MetaData = Field(description="Metadata for the prioritizer")
    rankedActionsMitigation: List[RankedAction] = Field(
        description="List of ranked mitigation actions"
    )
    rankedActionsAdaptation: List[RankedAction] = Field(
        description="List of ranked adaptation actions"
    )


class PrioritizerResponseBulk(BaseModel):
    prioritizerResponseList: List[PrioritizerResponse]
