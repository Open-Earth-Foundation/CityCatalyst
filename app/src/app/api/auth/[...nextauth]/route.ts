import { User } from "@/models/User";
import NextAuth from "next-auth";
import Credentials from "next-auth/providers/credentials";
import bcrypt from "bcrypt";

const handler = NextAuth({
  providers: [
    Credentials({
      name: "Credentials",
      credentials: {
        email: {
          label: "Email",
          type: "email",
          placeholder: "yourname@city.example",
        },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials): Promise<any> {
        if (!credentials) {
          return null;
        }
        let user: User | null = await User.findOne({
          where: { email: credentials.email },
        });
        if (!user) {
          return null;
        }
        if (
          user.passwordHash !== (await bcrypt.hash(credentials.password, 10))
        ) {
          return null;
        }
        return {
          id: user.userId,
          name: user.name,
          email: user.email,
          pictureUrl: user.pictureUrl,
          role: user.role,
        };
      },
    }),
  ],
});

export { handler as GET, handler as POST };
