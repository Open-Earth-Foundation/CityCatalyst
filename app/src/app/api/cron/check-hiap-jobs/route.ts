import { NextResponse } from "next/server";
import { db } from "@/models";
import { logger } from "@/services/logger";
import {
  HighImpactActionRankingStatus,
  ACTION_TYPES,
  LANGUAGES,
} from "@/util/types";
import { checkBulkActionRankingJob } from "@/backend/hiap/HiapService";
import { BulkHiapPrioritizationService } from "@/backend/hiap/BulkHiapPrioritizationService";
import { Op, QueryTypes } from "sequelize";

/**
 * Cron job endpoint to check HIAP job statuses and start next batches
 * Should be called every minute by Kubernetes CronJob
 * 

 * 1. Check all PENDING jobs - if completed, save results (mark SUCCESS/FAILURE)
 * 2. If NO PENDING jobs exist anywhere in the system:
 *    - Find ONE project with TO_DO rankings (oldest first, FIFO)
 *    - Start ONE batch for that project
 * 3. If ANY PENDING jobs exist, skip batch starting (wait for completion)
 * 
 * SECURITY: This endpoint is protected at the network level via Ingress.
 * The ingress blocks /api/cron/* paths from external access (see k8s/cc-ingress.yml).
 * Only internal Kubernetes services can call this endpoint.
 * 
 * @swagger
 * /api/cron/check-hiap-jobs:
 *   get:
 *     tags:
 *       - Cron
 *     summary: Check pending HIAP jobs and continue batch processing
 *     description: |
 *       Checks status of pending HIAP prioritization jobs and starts next batches.
 *       Also starts new batches for projects with TO_DO rankings but no active PENDING jobs.
 *     responses:
 *       200:
 *         description: Cron job completed
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 checkedJobs:
 *                   type: number
 *                 completedJobs:
 *                   type: number
 *                 startedBatches:
 *                   type: number
 */
export async function GET() {
  const startTime = Date.now();

  logger.info("üîÑ Cron job STARTED: Checking HIAP jobs");

  try {
    // Ensure database is initialized
    if (!db.initialized) {
      await db.initialize();
    }

    if (!db.sequelize) {
      throw new Error("Database not initialized");
    }

    // Step 1: Query for PENDING rankings grouped by unique jobId
    // Use DISTINCT to get unique jobIds efficiently in the database
    const pendingJobs = await db.sequelize.query<{
      jobId: string;
      type: ACTION_TYPES;
      langs: string[];
    }>(
      `
      SELECT DISTINCT ON ("job_id", "type") 
        "job_id" as "jobId",
        "type",
        "langs"
      FROM "public"."HighImpactActionRanking"
      WHERE "status" = :status
        AND "job_id" IS NOT NULL
      `,
      {
        replacements: { status: HighImpactActionRankingStatus.PENDING },
        type: QueryTypes.SELECT,
      },
    );

    logger.info(
      { pendingJobCount: pendingJobs.length },
      "Found pending HIAP jobs",
    );

    let completedJobs = 0;
    let startedBatches = 0;

    // Step 2: Check status for each unique jobId
    for (const job of pendingJobs) {
      try {
        const lang = (job.langs as any)[0] as LANGUAGES; // Get first language from array
        const isComplete = await checkBulkActionRankingJob(
          job.jobId,
          lang,
          job.type as ACTION_TYPES,
        );

        if (isComplete) {
          completedJobs++;

          // Step 3: Get projectId from one of the rankings
          const ranking = await db.models.HighImpactActionRanking.findOne({
            where: { jobId: job.jobId },
            include: [
              {
                model: db.models.Inventory,
                as: "inventory",
                required: true,
                include: [
                  {
                    model: db.models.City,
                    as: "city",
                    required: true,
                    attributes: ["projectId"],
                  },
                ],
              },
            ],
          });

          const projectId = (ranking as any)?.inventory?.city?.projectId;
          if (projectId) {
            // Step 4: Start next batch for this project if there are TO_DO rankings
            const nextBatch =
              await BulkHiapPrioritizationService.startNextBatch(
                projectId,
                job.type as ACTION_TYPES,
              );

            if (nextBatch.started) {
              startedBatches++;
              logger.info(
                {
                  projectId,
                  actionType: job.type,
                  batchSize: nextBatch.batchSize,
                  taskId: nextBatch.taskId,
                },
                "Started next batch",
              );
            }
          }
        }
      } catch (error: any) {
        logger.error(
          { jobId: job.jobId, error: error.message },
          "Error checking/processing HIAP job",
        );
        // Continue with other jobs even if one fails
      }
    }

    // Step 3: Start ONE batch if no PENDING jobs exist system-wide
    if (pendingJobs.length === 0) {
      logger.info(
        "No PENDING jobs system-wide. Checking for TO_DO rankings to start next batch.",
      );

      // Find ONE project with TO_DO rankings (oldest first for fairness)
      const nextProject = await db.sequelize.query<{
        projectId: string;
        type: ACTION_TYPES;
        oldestCreated: Date;
      }>(
        `
        SELECT c.project_id as "projectId", 
               har.type,
               MIN(har.created) as "oldestCreated"
        FROM "public"."HighImpactActionRanking" har
        JOIN "public"."Inventory" inv ON har.inventory_id = inv.inventory_id
        JOIN "public"."City" c ON inv.city_id = c.city_id
        WHERE har.status = :todoStatus
        GROUP BY c.project_id, har.type
        ORDER BY MIN(har.created) ASC
        LIMIT 1
        `,
        {
          replacements: {
            todoStatus: HighImpactActionRankingStatus.TO_DO,
          },
          type: QueryTypes.SELECT,
        },
      );

      if (nextProject.length > 0) {
        const project = nextProject[0];
        try {
          const nextBatch = await BulkHiapPrioritizationService.startNextBatch(
            project.projectId,
            project.type,
          );

          if (nextBatch.started) {
            startedBatches++;
            logger.info(
              {
                projectId: project.projectId,
                actionType: project.type,
                batchSize: nextBatch.batchSize,
                taskId: nextBatch.taskId,
                oldestCreated: project.oldestCreated,
              },
              "Started next batch",
            );
          }
        } catch (error: any) {
          logger.error(
            {
              projectId: project.projectId,
              actionType: project.type,
              error: error.message,
            },
            "Error starting batch",
          );
        }
      } else {
        logger.info("No TO_DO rankings found. All processing complete.");
      }
    } else {
      logger.info(
        {
          pendingJobCount: pendingJobs.length,
        },
        "PENDING jobs exist. Skipping Step 3 (enforce 1 batch at a time globally).",
      );
    }

    const duration = Date.now() - startTime;

    logger.info(
      {
        checkedJobs: pendingJobs.length,
        completedJobs,
        startedBatches,
        durationMs: duration,
      },
      "‚úÖ Cron job FINISHED successfully",
    );

    return NextResponse.json({
      checkedJobs: pendingJobs.length,
      completedJobs,
      startedBatches,
      durationMs: duration,
    });
  } catch (error: any) {
    const duration = Date.now() - startTime;
    logger.error(
      { error: error.message, stack: error.stack, durationMs: duration },
      "‚ùå Cron job FINISHED with error",
    );
    return NextResponse.json(
      { error: "Internal server error" },
      { status: 500 },
    );
  }
}

