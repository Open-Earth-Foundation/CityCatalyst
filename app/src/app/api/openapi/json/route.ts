/**
 * @swagger
 * /api/openapi/json:
 *   get:
 *     tags:
 *       - Documentation
 *     summary: Get OpenAPI specification in JSON format
 *     description: Returns the complete OpenAPI specification for the CityCatalyst API in JSON format.
 *     responses:
 *       200:
 *         description: OpenAPI specification returned successfully.
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               description: OpenAPI 3.0 specification
 *       500:
 *         description: Failed to generate OpenAPI specification.
 */
import { getApiDocs } from "@/lib/swagger";
import { NextResponse } from "next/server";
import { apiHandler } from "@/util/api";

export const GET = apiHandler(async (_req, { params, session }) => {
  try {
    const spec = await getApiDocs();
    
    return NextResponse.json(spec, {
      headers: {
        'Content-Type': 'application/json',
        'Cache-Control': 'public, max-age=3600', // Cache for 1 hour
      },
    });
  } catch (error) {
    console.error('Error generating OpenAPI spec:', error);
    return NextResponse.json(
      { 
        error: 'Failed to generate OpenAPI specification',
        details: error instanceof Error ? error.message : String(error)
      },
      { status: 500 }
    );
  }
});