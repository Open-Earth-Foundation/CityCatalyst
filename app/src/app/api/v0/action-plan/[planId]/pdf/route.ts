import { NextRequest, NextResponse } from "next/server";
import { apiHandler } from "@/util/api";
import PDFService, { ActionPlanData } from "@/backend/PDFService";
import createHttpError from "http-errors";

export const POST = apiHandler(async (req: NextRequest, { params }) => {
  try {
    const body = await req.json();
    const { planData, cityName, actionTitle, lng = "en" } = body;

    if (!planData || !cityName || !actionTitle) {
      throw createHttpError.BadRequest(
        "Missing required fields: planData, cityName, or actionTitle",
      );
    }

    // Validate planData has basic structure
    if (!planData.metadata && !planData.content) {
      throw createHttpError.BadRequest(
        "Invalid planData structure: must have metadata or content",
      );
    }

    // Generate PDF
    const pdfBuffer = await PDFService.generateActionPlanPDF(
      planData as ActionPlanData,
      cityName,
      actionTitle,
      lng,
    );

    // Return PDF as response
    return new NextResponse(pdfBuffer, {
      headers: {
        "Content-Type": "application/pdf",
        "Content-Disposition": `attachment; filename="action-plan-${actionTitle.replace(/[^a-zA-Z0-9]/g, "-")}.pdf"`,
        "Cache-Control": "no-cache",
      },
    });
  } catch (error: any) {
    console.error("PDF generation error:", error);

    if (error.statusCode) {
      throw error;
    }

    throw createHttpError.InternalServerError("Failed to generate PDF");
  }
});
