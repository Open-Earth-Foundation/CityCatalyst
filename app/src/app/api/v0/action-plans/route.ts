import { NextRequest, NextResponse } from "next/server";
import { apiHandler } from "@/util/api";
import ActionPlanService from "@/backend/ActionPlanService";
import { z } from "zod";
import createHttpError from "http-errors";

const getActionPlansSchema = z.object({
  cityId: z.string().optional(),
  language: z.string().optional(),
  actionId: z.string().optional(),
});

const createActionPlanSchema = z.object({
  actionId: z.string(),
  inventoryId: z.string().uuid(),
  hiActionRankingId: z.string().uuid().optional(),
  cityLocode: z.string(),
  actionName: z.string(),
  language: z.string(),
  planData: z.any(),
});

/**
 * @swagger
 * /api/v0/action-plans:
 *   get:
 *     summary: Get action plans
 *     description: Retrieve action plans with optional filtering by inventory ID, language, and action ID
 *     parameters:
 *       - in: query
 *         name: inventoryId
 *         schema:
 *           type: string
 *           format: uuid
 *         description: Filter by inventory ID
 *       - in: query
 *         name: language
 *         schema:
 *           type: string
 *         description: Filter by language
 *       - in: query
 *         name: actionId
 *         schema:
 *           type: string
 *         description: Filter by specific action ID
 *     responses:
 *       200:
 *         description: List of action plans
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 data:
 *                   type: array
 *                   items:
 *                     type: object
 */
export const GET = apiHandler(async (req: NextRequest) => {
  const url = new URL(req.url);
  const queryParams = Object.fromEntries(url.searchParams.entries());

  try {
    const { cityId, language, actionId } =
      getActionPlansSchema.parse(queryParams);

    const actionPlans = await ActionPlanService.getActionPlansByCityId(
      cityId,
      language,
      actionId,
    );

    return NextResponse.json({ data: actionPlans });
  } catch (error: any) {
    if (error instanceof z.ZodError) {
      throw createHttpError.BadRequest(
        `Invalid query parameters: ${error.message}`,
      );
    }
    throw error;
  }
});

/**
 * @swagger
 * /api/v0/action-plans:
 *   post:
 *     summary: Create a new action plan
 *     description: Create a new action plan with the provided data
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             required:
 *               - actionId
 *               - inventoryId
 *               - cityLocode
 *               - actionName
 *               - language
 *               - planData
 *             properties:
 *               actionId:
 *                 type: string
 *               inventoryId:
 *                 type: string
 *                 format: uuid
 *               hiActionRankingId:
 *                 type: string
 *                 format: uuid
 *               cityLocode:
 *                 type: string
 *               actionName:
 *                 type: string
 *               language:
 *                 type: string
 *               planData:
 *                 type: object
 *     responses:
 *       201:
 *         description: Action plan created successfully
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 data:
 *                   type: object
 */
export const POST = apiHandler(async (req: NextRequest, { session }) => {
  const body = await req.json();
  const validatedData = createActionPlanSchema.parse(body);

  const { actionPlan } = await ActionPlanService.upsertActionPlan({
    actionId: validatedData.actionId,
    highImpactActionRankedId: validatedData.hiActionRankingId,
    cityLocode: validatedData.cityLocode,
    actionName: validatedData.actionName,
    language: validatedData.language,
    planData: validatedData.planData,
    createdBy: session?.user?.id,
  });

  return NextResponse.json({ data: actionPlan }, { status: 201 });
});
