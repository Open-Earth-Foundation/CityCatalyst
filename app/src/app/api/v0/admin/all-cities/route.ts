/**
 * @swagger
 * /api/v0/admin/all-cities:
 *   get:
 *     tags:
 *       - Admin
 *     summary: List all cities with project and organization context.
 *     description: Returns all cities in the system including basic project and organization fields. Requires an admin session to succeed; non-admin users receive an authorization error. Use this to audit cities and their parent project/organization mapping.
 *     responses:
 *       200:
 *         description: List of cities wrapped in a data object.
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 data:
 *                   type: array
 *                   items:
 *                     type: object
 *                     properties:
 *                       cityId:
 *                         type: string
 *                         format: uuid
 *                       name:
 *                         type: string
 *                       locode:
 *                         type: string
 *                       project:
 *                         type: object
 *                         properties:
 *                           organizationId:
 *                             type: string
 *                             format: uuid
 *                           name:
 *                             type: string
 *                           cityCountLimit:
 *                             type: integer
 *                           organization:
 *                             type: object
 *                             properties:
 *                               organizationId:
 *                                 type: string
 *                                 format: uuid
 *                               name:
 *                                 type: string
 *                               contactEmail:
 *                                 type: string
 *                                 format: email
 *                     properties:
 *                       cityId:
 *                         type: string
 *                         format: uuid
 *                       name:
 *                         type: string
 *                       region:
 *                         type: string
 *                         nullable: true
 *                       country:
 *                         type: string
 *                         nullable: true
 *                       locode:
 *                         type: string
 *                         nullable: true
 *                       email:
 *                         type: string
 *                         format: email
 *                         nullable: true
 *             examples:
 *               example:
 *                 value:
 *                   data:
 *                     - cityId: "c9d8a3c2-1234-4c1a-9de1-6f3f25a2b111"
 *                       name: "Sample City"
 *                       locode: "US-XXX"
 *                       project:
 *                         organizationId: "b1c2d3e4-5678-4f90-aaaa-bbbbccccdddd"
 *                         name: "Project Alpha"
 *                         cityCountLimit: 25
 *                         organization:
 *                           organizationId: "0b6b1f1e-2222-4c33-9999-eeeeffff0000"
 *                           name: "Org Name"
 *                           contactEmail: "admin@example.org"
 *       401:
 *         description: Unauthorized.
 */
import { apiHandler } from "@/util/api";
import UserService from "@/backend/UserService";
import { db } from "@/models";
import { NextResponse } from "next/server";

export const GET = apiHandler(async (_req, { session }) => {
  UserService.validateIsAdmin(session);

  const cities = await db.models.City.findAll({
    include: [
      {
        model: db.models.Project,
        as: "project",
        attributes: ["organizationId", "name", "cityCountLimit"],
        include: [
          {
            model: db.models.Organization,
            as: "organization",
            attributes: ["organizationId", "name", "contactEmail"],
          },
        ],
      },
    ],
  });

  return NextResponse.json({ data: cities });
});
