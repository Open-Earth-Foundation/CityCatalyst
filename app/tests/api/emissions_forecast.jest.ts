import { afterAll, beforeAll, describe, expect, it, jest } from "@jest/globals";
import {
  baseInventory,
  growth_rates_response,
  inventoryId,
  inventoryValues as inventoryValuesData,
} from "./results.data";
import { GET as getResults } from "@/app/api/v0/inventory/[inventory]/results/emissions-forecast/route";
import { db } from "@/models";
import { randomUUID } from "node:crypto";
import { mockRequest, setupTests, testUserID } from "../helpers";
import {
  GlobalWarmingPotentialTypeEnum,
  InventoryTypeEnum,
} from "@/util/enums";
import { City } from "@/models/City";
import { Inventory } from "@/models/Inventory";

const locode = "XX_SUBCATEGORY_CITY";

// TODO UNSKIP when we can mock getGrowthRatesFromOC
describe.skip("Emissions Forecast API", () => {
  let city: City;
  let inventory: Inventory;
  beforeAll(async () => {
    setupTests();
    await db.initialize();
    city = await db.models.City.create({
      cityId: randomUUID(),
      locode,
    });
    await db.models.User.upsert({ userId: testUserID, name: "TEST_USER" });
    await city.addUser(testUserID);
    inventory = await db.models.Inventory.create({
      inventoryId: inventoryId,
      ...baseInventory,
      inventoryName: "EmissionsForecastInventory",
      cityId: city.cityId,
      inventoryType: InventoryTypeEnum.GPC_BASIC,
      globalWarmingPotentialType: GlobalWarmingPotentialTypeEnum.ar6,
    });

    await db.models.InventoryValue.bulkCreate(inventoryValuesData);
  });

  afterAll(async () => {
    await db.models.InventoryValue.destroy({ where: { inventoryId } });
    await db.models.Inventory.destroy({ where: { inventoryId } });
    await db.models.City.destroy({ where: { cityId: city.cityId } });
    if (db.sequelize) await db.sequelize.close();
  });

  it("should calculate projected emissions correctly", async () => {
    jest.mock("@/backend/OpenClimateService", () => {
      return {
        getGrowthRatesFromOC: jest
          .fn()
          .mockImplementation(() => growth_rates_response),
      };
    });
    const req = mockRequest();
    const result = await getResults(req, {
      params: { inventory: inventory.inventoryId },
    });

    expect(await result.json()).toMatchSnapshot();
  });

  it("should handle empty growth factors", async () => {
    jest.mock("@/backend/OpenClimateService", () => {
      return {
        getGrowthRatesFromOC: jest.fn().mockImplementation(() => undefined),
      };
    });
    const req = mockRequest();
    const result = await getResults(req, {
      params: { inventory: inventory.inventoryId },
    });

    expect(await result.json()).toEqual({
      data: {
        cluster: null,
        forecast: null,
        growthRates: null,
      },
    });
  });
});
