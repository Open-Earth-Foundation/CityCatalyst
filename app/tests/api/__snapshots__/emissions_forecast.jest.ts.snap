// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emissions Forecast API should calculate projected emissions correctly 1`] = `
{
  "data": {
    "cluster": {
      "description": {
        "en": "Emerging urban centers of industry (Non IPPU industry sectors)",
      },
      "id": 5,
    },
    "forecast": {
      "2025": {
        "I": "40399",
        "II": "22388",
        "III": "10581",
      },
      "2026": {
        "I": "42419",
        "II": "24403",
        "III": "11427",
      },
      "2027": {
        "I": "44964",
        "II": "25135",
        "III": "11770",
      },
      "2028": {
        "I": "47212",
        "II": "27397",
        "III": "12476",
      },
      "2029": {
        "I": "48156",
        "II": "30137",
        "III": "12601",
      },
      "2030": {
        "I": "51045",
        "II": "31945",
        "III": "13483",
      },
      "2031": {
        "I": "53087",
        "II": "34181",
        "III": "13618",
      },
      "2032": {
        "I": "56272",
        "II": "34865",
        "III": "14163",
      },
      "2033": {
        "I": "61336",
        "II": "37654",
        "III": "15438",
      },
      "2034": {
        "I": "61949",
        "II": "40290",
        "III": "15747",
      },
      "2035": {
        "I": "66285",
        "II": "42707",
        "III": "16062",
      },
      "2036": {
        "I": "68274",
        "II": "44842",
        "III": "17347",
      },
      "2037": {
        "I": "71688",
        "II": "46187",
        "III": "18388",
      },
      "2038": {
        "I": "78857",
        "II": "48496",
        "III": "19675",
      },
      "2039": {
        "I": "80434",
        "II": "48981",
        "III": "20069",
      },
      "2040": {
        "I": "85260",
        "II": "52899",
        "III": "21474",
      },
      "2041": {
        "I": "86965",
        "II": "54486",
        "III": "22977",
      },
      "2042": {
        "I": "93053",
        "II": "58300",
        "III": "23207",
      },
      "2043": {
        "I": "99567",
        "II": "58883",
        "III": "24831",
      },
      "2044": {
        "I": "103550",
        "II": "64182",
        "III": "26817",
      },
      "2045": {
        "I": "110799",
        "II": "70600",
        "III": "27622",
      },
      "2046": {
        "I": "118555",
        "II": "77660",
        "III": "28174",
      },
      "2047": {
        "I": "129225",
        "II": "78437",
        "III": "29301",
      },
      "2048": {
        "I": "135686",
        "II": "81574",
        "III": "30766",
      },
      "2049": {
        "I": "138400",
        "II": "84837",
        "III": "31381",
      },
      "2050": {
        "I": "143936",
        "II": "89079",
        "III": "32322",
      },
    },
    "growthRates": {
      "2024": {
        "I": 0.04,
        "II": 0.05,
        "III": 0.01,
        "IV": 0.05,
        "V": 0.08,
      },
      "2025": {
        "I": 0.04,
        "II": 0.05,
        "III": 0.01,
        "IV": 0.05,
        "V": 0.08,
      },
      "2026": {
        "I": 0.05,
        "II": 0.09,
        "III": 0.08,
        "IV": 0.06,
        "V": 0.05,
      },
      "2027": {
        "I": 0.06,
        "II": 0.03,
        "III": 0.03,
        "IV": 0.08,
        "V": 0.1,
      },
      "2028": {
        "I": 0.05,
        "II": 0.09,
        "III": 0.06,
        "IV": 0.02,
        "V": 0.05,
      },
      "2029": {
        "I": 0.02,
        "II": 0.1,
        "III": 0.01,
        "IV": 0.09,
        "V": 0.08,
      },
      "2030": {
        "I": 0.06,
        "II": 0.06,
        "III": 0.07,
        "IV": 0.04,
        "V": 0.03,
      },
      "2031": {
        "I": 0.04,
        "II": 0.07,
        "III": 0.01,
        "IV": 0.1,
        "V": 0.07,
      },
      "2032": {
        "I": 0.06,
        "II": 0.02,
        "III": 0.04,
        "IV": 0.04,
        "V": 0.06,
      },
      "2033": {
        "I": 0.09,
        "II": 0.08,
        "III": 0.09,
        "IV": 0.06,
        "V": 0.02,
      },
      "2034": {
        "I": 0.01,
        "II": 0.07,
        "III": 0.02,
        "IV": 0.06,
        "V": 0.09,
      },
      "2035": {
        "I": 0.07,
        "II": 0.06,
        "III": 0.02,
        "IV": 0.02,
        "V": 0.01,
      },
      "2036": {
        "I": 0.03,
        "II": 0.05,
        "III": 0.08,
        "IV": 0.06,
        "V": 0.08,
      },
      "2037": {
        "I": 0.05,
        "II": 0.03,
        "III": 0.06,
        "IV": 0.02,
        "V": 0.05,
      },
      "2038": {
        "I": 0.1,
        "II": 0.05,
        "III": 0.07,
        "IV": 0.06,
        "V": 0.02,
      },
      "2039": {
        "I": 0.02,
        "II": 0.01,
        "III": 0.02,
        "IV": 0.03,
        "V": 0.08,
      },
      "2040": {
        "I": 0.06,
        "II": 0.08,
        "III": 0.07,
        "IV": 0.04,
        "V": 0.03,
      },
      "2041": {
        "I": 0.02,
        "II": 0.03,
        "III": 0.07,
        "IV": 0.1,
        "V": 0.05,
      },
      "2042": {
        "I": 0.07,
        "II": 0.07,
        "III": 0.01,
        "IV": 0.1,
        "V": 0.09,
      },
      "2043": {
        "I": 0.07,
        "II": 0.01,
        "III": 0.07,
        "IV": 0.02,
        "V": 0.04,
      },
      "2044": {
        "I": 0.04,
        "II": 0.09,
        "III": 0.08,
        "IV": 0.04,
        "V": 0.09,
      },
      "2045": {
        "I": 0.07,
        "II": 0.1,
        "III": 0.03,
        "IV": 0.02,
        "V": 0.1,
      },
      "2046": {
        "I": 0.07,
        "II": 0.1,
        "III": 0.02,
        "IV": 0.06,
        "V": 0.07,
      },
      "2047": {
        "I": 0.09,
        "II": 0.01,
        "III": 0.04,
        "IV": 0.09,
        "V": 0.07,
      },
      "2048": {
        "I": 0.05,
        "II": 0.04,
        "III": 0.05,
        "IV": 0.03,
        "V": 0.02,
      },
      "2049": {
        "I": 0.02,
        "II": 0.04,
        "III": 0.02,
        "IV": 0.05,
        "V": 0.05,
      },
      "2050": {
        "I": 0.04,
        "II": 0.05,
        "III": 0.03,
        "IV": 0.08,
        "V": 0.01,
      },
    },
  },
}
`;
